docs/proposals/20200220-cluster-resource-set.md

```diff
@@ -0,0 +1,202 @@
+
+This is the CRD that is used to have a set of components that will be applied to clusters that match the label selector in it.
+
+The resources field is a list of secrets/configmaps in the same namespace. As a cluster label selector, any key-value pair works as long as the same key-value label is assigned to clusters that the addon will be applied to. The reason not to use a predefined label here is to allow matching with multiple ClusterResourceSet objects.
+
+*Sample ClusterResourceSet YAML*
+
+```yaml
+---
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: ClusterResourceSet
+metadata:
+ name: postcreate-conf
+ namespace: default
+spec:
+ mode: "ApplyOnce" / "Sync"
+ clusterSelector:
+   matchLabels:
+     postcreatelabelcni: calico
+ resources:
+   - name: calico-addon
+     kind: Secret
+   - name: network-policy-addon
+     kind: ConfigMap
+```
+
+Initially, the only mode supported will be "ApplyOnce" and it will be the default mode if no mode is provided. "Sync" mode will include reapplying the resources on resource hash change or periodically following an interval.
+Each item in the resources specifies a kind (must be either ConfigMap or Secret) and a name. Each referenced ConfigMap/Secret  contains yaml/json content as value. The key to that content is “value”.
+
+*Sample Secret Format*
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name:calico-addon
+type: Opaque
+stringData:
+  value: |-
+    kind: ConfigMap
+    apiVersion: v1
+    metadata:
+     name: calico-conf
+```
+
+The resources in ClusterResourceSet will be applied to matching clusters. 
+A configmap will be created in the management cluster to keep track of which resources are applied by ClusterResourceSet resources. There will be one configmap per workload cluster.
+
+Example:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: capi-crs-my-awesome-cluster
+  namespace: cluster-api-system
+data:
+  <ClusterResourceSet-name1>:   
+    <secret-name1>:
+      hash: <>
+      status: success
+      error: ""
+      timestamp: "2020-04-05T08:24:17Z"
+    <configmap-name1>:
+      hash: <>
+      status: failed
+      error: "some error"
+      timestamp: "2020-05-05T08:24:17Z"
+  <ClusterResourceSet-name2>:  
+    <secret-name2>:
+      hash: <>
+      status: success
+      error: ""
+      timestamp: "2020-04-05T08:24:17Z"
+  Status: InProgress/Completed
+```
+Status will be `Completed` when all matching ClusterResourceSet reconciles are completed for that cluster. In case of new resource addition to a matching ClusterResourceSet, Status becomes `InProgress`
+Also, the errors / overall progress will be tracked in the ClusterResourceSet’s status.
+
+### Risks and Mitigations
+
+Installing a component (such as CNI) using ClusterResourceSet that may later be managed by an addon operator for lifecycle management may require addon operators to discover and own those resources and reconciling should stop on those resources once this happens.
```

Comment:
I'm not sure this is a risk for CRS itself?

Reply:
One thing I can think of as a risk: when we introduce `Sync` mode, CNI operator may install a CNI to the cluster while ClusterResourceSet reconciles a same/different CNI. Not sure if this can fall into user error and not a risk.

As long as we don't support clean up of what ClusterResourceSet installs, this will be a risk on ClusterResourceSet side IMO.

Reply:
We will need to document that if you plan to have something else manage the lifecycle, you need to use mode=ApplyOnce.

Reply:
Removed this from the risks.

Added a note: 
`If ClusterResourceSet resources will be managed by an operator after they are applied by ClusterResourceSet controller, "ApplyOnce" mode must be used so that reconciliation on those resources can be delegated to the operator.
`